### ISSUE TITLE
The plugin can't deal with nested <> when editing Rust files.

### ISSUE THREAD

**Issue by b05902132** [LunarWatcher/auto-pairs#15]
**OS** (name + version; i.e. Linux Mint 20, Windows 10 2004, Macos Big Sur, etc.):
Archlinux
**What vim? (+ version)** 
Vim 8.2.2380

**Describe the bug**

When you insert nested <> (e.g. Arc<Mutex<...>>), there's only one closing > instead of two.

**Steps to reproduce**
Edit any .rs file, enter insert mode and type "Arc<Mutex<", Vim inserts "Arc<Mutex<>" instead of "Arc<Mutex<>>".

**Comment by LunarWatcher** [comment]
No repro in any other files but Rust files. Mildly sus it's down to [the Rust plugin remapping stuff](https://github.com/rust-lang/rust.vim/blob/96e79e397126be1a64fb53d8e3656842fe1a4532/ftplugin/rust.vim#L77-L79). Are you using that plugin? If yes, can you `let g:rust_keep_autopairs_default = 1` for science? Might also be a part of the polyglot plugin, if you haven't explicitly defined a Rust plugin.

I have absolutely no clue why it only inserts one though -- can't repro that locally in or out of Rust files. What happens if you `:verbose imap <`?

**Comment by b05902132** [comment]
> No repro in any other files but Rust files. Mildly sus it's down to [the Rust plugin remapping stuff](https://github.com/rust-lang/rust.vim/blob/96e79e397126be1a64fb53d8e3656842fe1a4532/ftplugin/rust.vim#L77-L79). Are you using that plugin? If yes, can you `let g:rust_keep_autopairs_default = 1` for science? Might also be a part of the polyglot plugin, if you haven't explicitly defined a Rust plugin.
I use neither of those.
> 
> I have absolutely no clue why it only inserts one though -- can't repro that locally in or out of Rust files. What happens if you `:verbose imap <`?
It shows `i  <           *@<C-R>=autopairs#AutoPairsInsert('<')<CR>`.

I believe it's due to language specific pair. I just edited my HTML pair to be `{\w\zs<': '>'}`, and this bug also shows up.

**Comment by LunarWatcher** [comment]
Managed to repro with that one specifically. You could set that one to be specifically for HTML though (see `g:AutoPairsLanguagePairs` and temporarily `autopairs#AutoPairsScriptInit`). Gonna look into this though - the regex itself matches both, so I'm not entirely sure what happens here. Might be a borked balance check somewhere

**Comment by LunarWatcher** [comment]
Try updating from the master branch. It was a borked balance check :')

TL;DR: count doesn't process regex, so your pair registered a close without an open, which tripped the balance check.

**Comment by b05902132** [comment]
It's working! Thank you.

**Comment by LunarWatcher** [comment]
I don't _think_ it should cause issues, but not using \V in the regex method might come back to bite me. We'll see, didn't cause problems in testing - if it causes problems, feel free to open another issue or leave a comment here.

**Comment by b05902132** [comment]
> No repro in any other files but Rust files. Mildly sus it's down to [the Rust plugin remapping stuff](https://github.com/rust-lang/rust.vim/blob/96e79e397126be1a64fb53d8e3656842fe1a4532/ftplugin/rust.vim#L77-L79). Are you using that plugin? If yes, can you `let g:rust_keep_autopairs_default = 1` for science? Might also be a part of the polyglot plugin, if you haven't explicitly defined a Rust plugin.

By the way, I just installed that plugin, and it disables auto pair for `<>`. `let g:rust_keep_autopairs_default` enables it again.

Think you may want to know.

**Comment by LunarWatcher** [comment]
Yeah, it forcibly overrides `b:AutoPairs`. The variable itself is largely meant for when you'd want a specific set of pairs for a buffer (i.e. for a specific filetype, as in a complete override instead of just custom additions, unless `autopairs#AutoPairsDefine` is used to add pairs). 

Setting it that way is kinda counter-productive (in my opinion, to be clear), as it completely overrides user config with how this plugin is set up. Might have to find an alternate system that lets plugins and user options coexist, but I have no ideas for that atm. `autopairs#AutoPairsDefine` might already be a good choice, as it enables pair removal in the second argument (which is a vararg), but gonna have to deal with that later. Compatibility is also a bit of a problem -- I don't think I currently have enough "justification" or whatever to alter other plugins to facilitate this one. We'll see, future me's problem and whatnot :p
